packages/@aws-cdk/aws-codepipeline-api/lib/action.ts

```diff
@@ -0,0 +1,305 @@
+import events = require('@aws-cdk/aws-events');
+import iam = require('@aws-cdk/aws-iam');
+import cdk = require('@aws-cdk/cdk');
+import { Artifact } from './artifact';
+import validation = require('./validation');
+
+export enum ActionCategory {
+    Source = 'Source',
+    Build = 'Build',
+    Test = 'Test',
+    Approval = 'Approval',
+    Deploy = 'Deploy',
+    Invoke = 'Invoke'
+}
+
+/**
+ * Specifies the constraints on the number of input and output
+ * artifacts an action can have.
+ *
+ * The constraints for each action type are documented on the
+ * {@link https://docs.aws.amazon.com/codepipeline/latest/userguide/reference-pipeline-structure.html Pipeline Structure Reference} page.
+ */
+export interface ActionArtifactBounds {
+    readonly minInputs: number;
+    readonly maxInputs: number;
+    readonly minOutputs: number;
+    readonly maxOutputs: number;
+}
+
+export function defaultBounds(): ActionArtifactBounds {
+    return {
+        minInputs: 0,
+        maxInputs: 5,
+        minOutputs: 0,
+        maxOutputs: 5
+    };
+}
+
+/**
+ * The abstract interface of a Pipeline Stage that is used by Actions.
+ */
+export interface IStage {
+    /**
+     * The physical, human-readable name of this Pipeline Stage.
+     */
+    readonly name: string;
+
+    /**
+     * The ARN of the Pipeline.
+     */
+    readonly pipelineArn: cdk.Arn;
+
+    /**
+     * The service Role of the Pipeline.
+     */
+    readonly pipelineRole: iam.Role;
+
+    /**
+     * Grants read & write permissions to the Pipeline's S3 Bucket to the given Identity.
+     *
+     * @param identity the IAM Identity to grant the permissions to
+     */
+    grantPipelineBucketReadWriteTo(identity?: iam.IPrincipal): void;
+
+    /**
+     * Adds an Action to this Stage.
+     * This is an internal operation -
+     * an Action is added to a Stage when it's constructed,
+     * so there's no need to call this method explicitly.
+     *
+     * @param action the Action to add to this Stage
+     */
+    _addAction(action: Action): void;
+}
+
+/**
+ * Common properties shared by all Actions.
+ */
+export interface AbstractActionProps {
```

Comment:
I still think this is redundant, but if you insist, I'd call this `CommonActionProps`.

Reply:
👍🏻